src/main/java/picard/illumina/parser/readers/CbclReader.java

```diff
@@ -180,126 +247,124 @@ public void close() {
 
         for (int read = 0; read < outputLengths.length; read++) {
             for (int cycle = 0; cycle < outputLengths[read]; cycle++) {
-                try {
-                    CycleData currentCycleData = cycleData[totalCycleCount];
-                    TileData currentTileData = currentCycleData.tileInfo[currentTile[totalCycleCount]];
-                    try {
-                        if (cachedTile[totalCycleCount] == null) {
-                            if (!cachedFilter.containsKey(currentTileData.tileNum)) {
-                                cacheFilter(currentTileData);
-                            }
-                            cacheTile(totalCycleCount, currentTileData, currentCycleData);
-                        }
-                    } catch (IOException e) {
-                        // when logging the error, increment cycle by 1, since totalCycleCount is zero-indexed but Illumina directories are 1-indexed.
-                        throw new IOException(String.format("Error while reading from BCL file for cycle %d. Offending file on disk is %s",
-                                (totalCycleCount + 1), this.streamFiles[totalCycleCount].getAbsolutePath()), e);
-                    }
+                CycleData currentCycleData = cycleData[totalCycleCount];
 
-                    if (!cachedTile[totalCycleCount].hasRemaining()) {
-                        //on to the next tile
-                        currentTile[totalCycleCount]++;
-                        //if past the last tile then return
-                        if (currentTile[totalCycleCount] > currentCycleData.tileInfo.length - 1) {
-                            return;
-                        }
-                        currentTileData = currentCycleData.tileInfo[currentTile[totalCycleCount]];
-                        if (!cachedFilter.containsKey(currentTileData.tileNum)) {
-                            cacheFilter(currentTileData);
-                        }
-                        cacheTile(totalCycleCount, currentTileData, currentCycleData);
-                    }
-
-                    int singleByte = cachedTile[totalCycleCount].get();
+                if (cachedTilePosition[totalCycleCount] >= cachedTile[totalCycleCount].length) {
+                    // end of tile
+                    return;
+                }
 
-                    decodeQualityBinnedBasecall(data, read, cycle, singleByte, currentCycleData);
+                int singleByte = cachedTile[totalCycleCount][cachedTilePosition[totalCycleCount]++];
 
-                    totalCycleCount++;
-                } catch (final IOException ioe) {
-                    throw new RuntimeIOException(ioe);
-                }
+                decodeQualityBinnedBasecall(data, read, cycle, singleByte, currentCycleData);
 
+                totalCycleCount++;
             }
         }
-        this.queue.add(data);
+        data.setPositionInfo(positionInfoIterator.next());
```

Comment:
This part of the code is confusing to me. I was trying to figure out why this was set here instead of being passed in as an argument to the `CbclData` constructor. I see that this iterator is updated both in `advance()` and in `cacheFilterAndLocs()` but it wasn't obvious to me if one of these was called from the other or not.